### PR TITLE
added browser entry point to package.json and excluded backend-only drivers

### PR DESCRIPTION
**Pull request by chriswep**
There is a configuration option for `package.json` to instruct bundlers like `webpack` to use a different entry point in browser environments without any further configuration needed:  https://github.com/defunctzombie/package-browser-field-spec

Using this, we can additionally replace backend-only drivers by dummy classes to reduce the bundle size. Disabling Mongo, Postgres, SqlServer, Mysql and Oracle reduces the bundle size by one third.

With this change, `typeorm` can be imported in browser environments with zero additional configuration.

### CODE REVIEW

**Comment by vlapo** [comment]
Fixes https://github.com/typeorm/typeorm/issues/3583

**Comment by pleerock** [comment]
what about query runners? Are they excluded this way?

**Comment by chriswep** [comment]
> what about query runners? Are they excluded this way?

i checked with the weback BundleAnalyzer: somehow the mongodb query runner survived, however the other excluded drivers are completely gone.

@pleerock as mentioned in #3583 - concerning the duplicate classes and IDE suggestions: would it be a compromise to add jsdoc warnings to the classes? that should prevent most cases of importing the wrong class. maybe there is also a possibility to ignore the file via `editorconfig`?

Anyway, if want to have driver-replacements out of this pull request i can remove them and just leave the new browser entry point in `package.json`.

**Comment by chriswep** [comment]
> > what about query runners? Are they excluded this way?
> 
> i checked with the weback BundleAnalyzer: somehow the mongodb query runner survived, however the other excluded drivers are completely gone.

seems like MongoDb has a special spot in the codebase (for historic reasons?) - apart from the mongodb query runner also the mongodb entity manger and repository are imported from outside the driver (this doesn't happen for the other drivers). Anyway, those files can be replaced by the same method via `package.json` if we wanted to (bringing down the bundle size a bit more).

**Comment by pleerock** [comment]
> would it be a compromise to add jsdoc warnings to the classes? that should prevent most cases of importing the wrong class. maybe there is also a possibility to ignore the file via `editorconfig`?

I don't think editorconfig has such option. jsdoc also wouldn't help. The only option I know is to have this file in format other than `.ts`, and create a separate build configuration that will inject this file only in the build process (we already do it this way, see BrowserPlatformTools file).

Yes, we need to completely strip out mongo as well.

**Comment by chriswep** [comment]
> The only option I know is to have this file in format other than `.ts`, and create a separate build configuration that will inject this file only in the build process (we already do it this way, see BrowserPlatformTools file).

done
 
> Yes, we need to completely strip out mongo as well.

done

@pleerock any other concerns?

**Comment by pleerock** [comment]
Hope you won't have me for the solution I suggested. Hope we can came up to something better in the future.

Thank you very much for your contribution!

**Comment by timoschwarzer** [comment]
There's one problem with this. When using Electron in combination with Vue CLI (which uses Webpack internally), TypeORM stops working with this patch. Version 0.2.13 worked just fine but versions since 0.2.14 don't work anymore. What would be the best way to fix this? :smiley:

**Comment by Kononnable** [comment]
A lot of libs(electron, vue, webpack), so it might be hard to find a problem. It probably doesn't work out with some of webpack configuration, but with which option? What error does it throw? We should probably create new issue for this - it would be hard to follow something in already merged PR.

**Comment by chriswep** [comment]
@timoschwarzer you can tell webpack to use another entry point (i.e. main, not browser, which is the default)

**Comment by timoschwarzer** [comment]
@Kononnable you're right. For more details see my newly created issue. (tl;dr: target=node does not work with Electron)

**Comment by timoschwarzer** [comment]
@chriswep Do you mind telling me how to do that? :)
